presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,269 @@
+        requireNonNull(sources, "sources is null");
+        requireNonNull(outputBuffers, "outputBuffers is null");
+        requireNonNull(httpClient, "httpClient is null");
+
+        try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
+            this.taskId = taskId;
+            this.session = session;
+            this.planFragment = planFragment;
+            this.outputBuffers.set(outputBuffers);
+            this.httpClient = httpClient;
+            this.tableWriteInfo = tableWriteInfo;
+            this.stats = stats;
+            this.sources = sources;
+            this.executor = executor;
+            this.updateScheduledExecutor = updateScheduledExecutor;
+            List<BufferInfo> bufferStates = outputBuffers.getBuffers()
+                    .keySet().stream()
+                    .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
+                    .collect(toImmutableList());
+            this.taskInfo = createInitialTask(taskId, location, bufferStates, new TaskStats(DateTime.now(), null), nodeId);
+            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
+            this.planFragmentCodec = planFragmentCodec;
+            this.taskInfoCodec = taskInfoCodec;
+        }
+    }
+
+    public TaskInfo getTaskInfo()
+    {
+        return taskInfo;
+    }
+
+    public void start()
+    {
+        sendUpdate();
+        scheduleGetTaskInfo();
+    }
+
+    private synchronized void sendUpdate()
+    {
+        Optional<byte[]> fragment = Optional.of(planFragment.toBytes(planFragmentCodec));
+        Optional<TableWriteInfo> writeInfo = Optional.of(tableWriteInfo);
+        TaskUpdateRequest updateRequest = new TaskUpdateRequest(
+                session.toSessionRepresentation(),
+                session.getIdentity().getExtraCredentials(),
+                fragment,
+                sources,
+                outputBuffers.get(),
+                writeInfo);
+        byte[] taskUpdateRequestJson = taskUpdateRequestCodec.toBytes(updateRequest);
+
+        HttpUriBuilder uriBuilder = uriBuilderFrom(taskInfo.getTaskStatus().getSelf());
+        Request request = setContentTypeHeaders(false, preparePost())
+                .setUri(uriBuilder.build())
+                .setBodyGenerator(createStaticBodyGenerator(taskUpdateRequestJson))
+                .build();
+
+        ResponseHandler responseHandler = createAdaptingJsonResponseHandler(taskInfoCodec);
+        ListenableFuture<BaseResponse<TaskInfo>> future = httpClient.executeAsync(request, responseHandler);
+        FutureCallback callback = new SimpleHttpResponseHandler<>(new UpdateResponseHandler(), request.getUri(), stats.getHttpResponseStats(), NATIVE_EXECUTION_TASK_ERROR);
+
+        Futures.addCallback(
+                future,
+                callback,
+                executor);
+    }
+
+    private synchronized void scheduleGetTaskInfo()
```

Comment:
Why does this method need to be synchronized?

Reply:
I see. The synchronization is to protect taskInfo. You want to protect taskInfo. However I think that is not protected because taskInfo is updated in handler class on another thread (when future executes callback). I'm not sure how PrestoDB handles the situation in this case. In Velox/Prestissimo we have a dedicated mutex to protect that specific field. There might be a similar approach (or even simpler as it's java) @highker

Reply:
good point. one approach I can think of is 
add a synchronized function updateTaskInfo()
will update the PR